In [1476]:
import pandas as pd
pd.set_option('display.max_rows', None)  # 显示所有行
pd.set_option('display.max_columns', None)  # 显示所有列

In [1477]:
file_path = 'depth.csv'
df = pd.read_csv(file_path)

# df = df.drop(columns=['pbid'])
df = df[df['depth'] >= 20]
df = df[df['num_params'] >= 3]
grouped_df = df.groupby(['pkid', 'layers', 'method'], as_index=False).agg({
    "depth": 'mean',
    # 'culled_depth': 'mean',
    'num_params': 'mean',
})
# values = ["depth", "culled_depth"]
values = ["depth", 'num_params']

## 分组并把组作为索引
pivot_df = grouped_df.pivot(index =['pkid'], columns='method', values=values)

method_order = ['HeaSolver', 'PenaltySolver', 'ChocoSolver', 'QTOSimplifyDiscardSolver']
pivot_df = pivot_df.reindex(columns=pd.MultiIndex.from_product([values, method_order]))

pivot_df


depth                                                    num_params  \
     HeaSolver PenaltySolver ChocoSolver QTOSimplifyDiscardSolver  HeaSolver   
pkid                                                                           
0         46.0          87.0       507.0               102.000000       90.0   
1         86.0         153.0      1532.0               306.000000      210.0   
2        121.0         174.0      2688.0              1312.000000      315.0   
3        156.0         207.0      3848.0              2910.800000      420.0   
4         76.0         259.8      1878.5               176.000000      180.0   
5         91.0         309.0      1678.5               193.000000      225.0   
6        136.0         785.4      3656.8               312.100000      360.0   
7        156.0         833.7      3263.4               256.444444      420.0   
8         56.0         175.0      1328.2               299.000000      120.0   
9        106.0         282.3      4395.3              1126.400000      270.0   
10       136.0         356.2      6137.0              3056.300000      360.0   
11       151.0         404.5      7024.4              3308.300000      405.0   
12        51.0         110.0      1062.0               215.000000      105.0   
13        86.0         170.1      3217.0               613.100000      210.0   
14       106.0         191.0      4562.0               756.800000      270.0   
15       151.0         240.0      7687.0              1815.800000      405.0   
16        61.0         221.3      1304.4               234.444444      135.0   
17        76.0         292.6      1483.6               264.000000      180.0   
18       116.0         489.8      2194.1              1084.300000      300.0   
19       141.0         627.7      2906.2              1977.000000      375.0   

                                                         
     PenaltySolver ChocoSolver QTOSimplifyDiscardSolver  
pkid                                                     
0             10.0        10.0                 3.000000  
1             10.0        10.0                 5.000000  
2             10.0        10.0                21.000000  
3             10.0        10.0                52.000000  
4             10.0        10.0                 4.000000  
5             10.0        10.0                 3.000000  
6             10.0        10.0                 6.200000  
7             10.0        10.0                 5.222222  
8             10.0        10.0                 3.500000  
9             10.0        10.0                13.100000  
10            10.0        10.0                35.400000  
11            10.0        10.0                38.300000  
12            10.0        10.0                 3.500000  
13            10.0        10.0                10.000000  
14            10.0        10.0                11.200000  
15            10.0        10.0                24.300000  
16            10.0        10.0                 7.555556  
17            10.0        10.0                12.000000  
18            10.0        10.0                51.000000  
19            10.0        10.0               116.700000

In [1478]:
culled_depth_qtosimplify_discard = pivot_df['depth']['QTOSimplifyDiscardSolver']
num_params_qtosimplify_discard = pivot_df['num_params']['QTOSimplifyDiscardSolver']

# 计算新的列并将其添加到 `pivot_df`
pivot_df[('culled_depth_per_param', 'QTOSimplifyDiscardSolver')] = culled_depth_qtosimplify_discard / num_params_qtosimplify_discard
pivot_df

depth                                                    num_params  \
     HeaSolver PenaltySolver ChocoSolver QTOSimplifyDiscardSolver  HeaSolver   
pkid                                                                           
0         46.0          87.0       507.0               102.000000       90.0   
1         86.0         153.0      1532.0               306.000000      210.0   
2        121.0         174.0      2688.0              1312.000000      315.0   
3        156.0         207.0      3848.0              2910.800000      420.0   
4         76.0         259.8      1878.5               176.000000      180.0   
5         91.0         309.0      1678.5               193.000000      225.0   
6        136.0         785.4      3656.8               312.100000      360.0   
7        156.0         833.7      3263.4               256.444444      420.0   
8         56.0         175.0      1328.2               299.000000      120.0   
9        106.0         282.3      4395.3              1126.400000      270.0   
10       136.0         356.2      6137.0              3056.300000      360.0   
11       151.0         404.5      7024.4              3308.300000      405.0   
12        51.0         110.0      1062.0               215.000000      105.0   
13        86.0         170.1      3217.0               613.100000      210.0   
14       106.0         191.0      4562.0               756.800000      270.0   
15       151.0         240.0      7687.0              1815.800000      405.0   
16        61.0         221.3      1304.4               234.444444      135.0   
17        76.0         292.6      1483.6               264.000000      180.0   
18       116.0         489.8      2194.1              1084.300000      300.0   
19       141.0         627.7      2906.2              1977.000000      375.0   

                                                         \
     PenaltySolver ChocoSolver QTOSimplifyDiscardSolver   
pkid                                                      
0             10.0        10.0                 3.000000   
1             10.0        10.0                 5.000000   
2             10.0        10.0                21.000000   
3             10.0        10.0                52.000000   
4             10.0        10.0                 4.000000   
5             10.0        10.0                 3.000000   
6             10.0        10.0                 6.200000   
7             10.0        10.0                 5.222222   
8             10.0        10.0                 3.500000   
9             10.0        10.0                13.100000   
10            10.0        10.0                35.400000   
11            10.0        10.0                38.300000   
12            10.0        10.0                 3.500000   
13            10.0        10.0                10.000000   
14            10.0        10.0                11.200000   
15            10.0        10.0                24.300000   
16            10.0        10.0                 7.555556   
17            10.0        10.0                12.000000   
18            10.0        10.0                51.000000   
19            10.0        10.0               116.700000   

       culled_depth_per_param  
     QTOSimplifyDiscardSolver  
pkid                           
0                   34.000000  
1                   61.200000  
2                   62.476190  
3                   55.976923  
4                   44.000000  
5                   64.333333  
6                   50.338710  
7                   49.106383  
8                   85.428571  
9                   85.984733  
10                  86.336158  
11                  86.378590  
12                  61.428571  
13                  61.310000  
14                  67.571429  
15                  74.724280  
16                  31.029412  
17                  22.000000  
18                  21.260784  
19                  16.940874

In [1479]:
mean_values_per_column = pivot_df.mean()
print(mean_values_per_column)

depth                   HeaSolver                    105.250000
                        PenaltySolver                318.470000
                        ChocoSolver                 3117.670000
                        QTOSimplifyDiscardSolver    1015.939444
num_params              HeaSolver                    267.750000
                        PenaltySolver                 10.000000
                        ChocoSolver                   10.000000
                        QTOSimplifyDiscardSolver      21.298889
culled_depth_per_param  QTOSimplifyDiscardSolver      56.091247
dtype: float64


In [1480]:
# 定义格式化函数
import numpy as np
# def format_value(val):
#     if np.isnan(val):  # 如果是NaN值则返回空字符
#         return "NaN"
#     elif val < 100:  # 如果值小于100，保留3位有效数字
#         return f"{val:.2f}"
#     else:  # 如果值大于等于100，不保留小数点
#         return f"{val:.0f}"

# 按行遍历输出，每个值用 & 分隔
for row in pivot_df.itertuples(index=False):
    formatted_row = [f"{x:.0f}" for x in row]
    print(" & ".join(formatted_row))

46 & 87 & 507 & 102 & 90 & 10 & 10 & 3 & 34
86 & 153 & 1532 & 306 & 210 & 10 & 10 & 5 & 61
121 & 174 & 2688 & 1312 & 315 & 10 & 10 & 21 & 62
156 & 207 & 3848 & 2911 & 420 & 10 & 10 & 52 & 56
76 & 260 & 1878 & 176 & 180 & 10 & 10 & 4 & 44
91 & 309 & 1678 & 193 & 225 & 10 & 10 & 3 & 64
136 & 785 & 3657 & 312 & 360 & 10 & 10 & 6 & 50
156 & 834 & 3263 & 256 & 420 & 10 & 10 & 5 & 49
56 & 175 & 1328 & 299 & 120 & 10 & 10 & 4 & 85
106 & 282 & 4395 & 1126 & 270 & 10 & 10 & 13 & 86
136 & 356 & 6137 & 3056 & 360 & 10 & 10 & 35 & 86
151 & 404 & 7024 & 3308 & 405 & 10 & 10 & 38 & 86
51 & 110 & 1062 & 215 & 105 & 10 & 10 & 4 & 61
86 & 170 & 3217 & 613 & 210 & 10 & 10 & 10 & 61
106 & 191 & 4562 & 757 & 270 & 10 & 10 & 11 & 68
151 & 240 & 7687 & 1816 & 405 & 10 & 10 & 24 & 75
61 & 221 & 1304 & 234 & 135 & 10 & 10 & 8 & 31
76 & 293 & 1484 & 264 & 180 & 10 & 10 & 12 & 22
116 & 490 & 2194 & 1084 & 300 & 10 & 10 & 51 & 21
141 & 628 & 2906 & 1977 & 375 & 10 & 10 & 117 & 17


In [1481]:
# 选取 'depth' 下的所有列，并剔除 'NewSolver' 和 'QTOSimplifySolver'
depth_df = pivot_df['depth'].drop(columns=['NewSolver', 'QTOSimplifySolver'], errors='ignore')

# 选取 'culled_depth_per_param' 下的 'QTOSimplifyDiscardSolver' 列
qtosimplify_discard_df = pivot_df['culled_depth_per_param'][['QTOSimplifyDiscardSolver']]

# 合并处理后的 'depth' 列与 'QTOSimplifyDiscardSolver' 列，并指定后缀避免冲突
culled_depth_df = depth_df.join(qtosimplify_discard_df, lsuffix='_depth', rsuffix='_culled')

# 按行遍历输出值，每个值用 & 分隔
for row in culled_depth_df.itertuples(index=False):
    formatted_row = [f"{x:5.0f}" for x in row]
    print(" & ".join(formatted_row))


   46 &    87 &   507 &   102 &    34
   86 &   153 &  1532 &   306 &    61
  121 &   174 &  2688 &  1312 &    62
  156 &   207 &  3848 &  2911 &    56
   76 &   260 &  1878 &   176 &    44
   91 &   309 &  1678 &   193 &    64
  136 &   785 &  3657 &   312 &    50
  156 &   834 &  3263 &   256 &    49
   56 &   175 &  1328 &   299 &    85
  106 &   282 &  4395 &  1126 &    86
  136 &   356 &  6137 &  3056 &    86
  151 &   404 &  7024 &  3308 &    86
   51 &   110 &  1062 &   215 &    61
   86 &   170 &  3217 &   613 &    61
  106 &   191 &  4562 &   757 &    68
  151 &   240 &  7687 &  1816 &    75
   61 &   221 &  1304 &   234 &    31
   76 &   293 &  1484 &   264 &    22
  116 &   490 &  2194 &  1084 &    21
  141 &   628 &  2906 &  1977 &    17


In [1482]:
# 选取所需的列
depth_qto_discard = pivot_df['depth']['QTOSimplifyDiscardSolver']
depth_choco = pivot_df['depth']['ChocoSolver']
culled_depth_qto_discard = pivot_df['culled_depth_per_param']['QTOSimplifyDiscardSolver']
# 方法1：每行相除取平均
# depth 下 ChocoSolver / QTOSimplifyDiscardSolver 的倍数
row_wise_ratio_depth_choco_qto_discard = (depth_choco / depth_qto_discard).mean()

# culled_depth_per_param 下 ChocoSolver / QTOSimplifyDiscardSolver 的倍数
row_wise_ratio_culled_depth_choco_qto_discard = (depth_choco / culled_depth_qto_discard).mean()

# 方法2：整列取平均后相除
# depth 下 ChocoSolver / QTOSimplifyDiscardSolver 的整列平均的倍数
col_avg_ratio_depth_choco_qto_discard = depth_choco.mean() / depth_qto_discard.mean()

# culled_depth_per_param 下 ChocoSolver / QTOSimplifyDiscardSolver 的整列平均的倍数
col_avg_ratio_culled_depth_choco_qto_discard = depth_choco.mean() / culled_depth_qto_discard.mean()

# 输出结果
print(f"每行相除取平均 - depth 下 ChocoSolver / QTOSimplifyDiscardSolver 的倍数: {row_wise_ratio_depth_choco_qto_discard:.2f}")
print(f"每行相除取平均 - depth 下 ChocoSolver / culled_depth_per_param 下 QTOSimplifyDiscardSolver 的倍数: {row_wise_ratio_culled_depth_choco_qto_discard:.2f}")
print(f"整列取平均后相除 - depth 下 ChocoSolver / QTOSimplifyDiscardSolver 的倍数: {col_avg_ratio_depth_choco_qto_discard:.2f}")
print(f"整列取平均后相除 - depth 下 ChocoSolver / culled_depth_per_param 下 QTOSimplifyDiscardSolver 的倍数: {col_avg_ratio_culled_depth_choco_qto_discard:.2f}")


每行相除取平均 - depth 下 ChocoSolver / QTOSimplifyDiscardSolver 的倍数: 5.24
每行相除取平均 - depth 下 ChocoSolver / culled_depth_per_param 下 QTOSimplifyDiscardSolver 的倍数: 60.15
整列取平均后相除 - depth 下 ChocoSolver / QTOSimplifyDiscardSolver 的倍数: 3.07
整列取平均后相除 - depth 下 ChocoSolver / culled_depth_per_param 下 QTOSimplifyDiscardSolver 的倍数: 55.58


\multirow{4}{*}{\makecell[l]{\textbf{FLP}\\~\cite{melo2009facilitylocation}}} 
& \FLbcmk{2}{1} & 6  & 3  &    46 &    87 &   507 &   102 &    34 \\
& \FLbcmk{3}{2} & 15 & 8  &    86 &   153 &  1532 &   306 &    61 \\
& \FLbcmk{3}{3} & 21 & 12 &   121 &   174 &  2688 &  1313 &    62 \\
& \FLbcmk{4}{3} & 28 & 15 &   156 &   207 &  3848 &  3135 &    57 \\
\hline

\multirow{4}{*}{\makecell[l]{\textbf{GCP}\\~\cite{graphcoloring}}}
& \GCbcmk{3}{1} & 12 & 6  &    76 &   259 &  1879 &   176 &    44 \\
& \GCbcmk{3}{2} & 15 & 9  &    91 &   310 &  1697 &   140 &    60 \\
& \GCbcmk{4}{2} & 24 & 12 &   136 &   791 &  3515 &   313 &    48 \\
& \GCbcmk{4}{3} & 28 & 16 &   156 &   840 &  3391 &   279 &    52 \\
\hline

\multirow{4}{*}{\makecell[l]{\textbf{SCP}\\~\cite{tsp}}} 
& \SCbcmk{4}{4} & 9  & 4 &   56 &   175 &  1324 &   299 &    85 \\
& \SCbcmk{5}{5} & 12 & 5 &  106 &   284 &  4398 &  1210 &    86 \\
& \SCbcmk{7}{7} & 20 & 7 &  136 &   363 &  6150 &  3274 &    86 \\
& \SCbcmk{8}{8} & 25 & 8 &  151 &   404 &  7021 &  4643 &    87 \\
\hline

\multirow{4}{*}{\makecell[l]{\textbf{KPP}\\~\cite{kpartition}}} 
& \KPbcmk{4}{3}{2} & 8  & 6  &    51 &   110 &  1058 &   214 &    61 \\
& \KPbcmk{6}{5}{3} & 18 & 9  &    86 &   171 &  3189 &   642 &    61 \\
& \KPbcmk{8}{7}{3} & 24 & 11 &   106 &   193 &  4470 &  1082 &    69 \\
& \KPbcmk{9}{8}{3} & 27 & 12 &   151 &   239 &  7623 &  2455 &    76 \\
\hline

\multirow{4}{*}{\makecell[l]{\textbf{JSP}\\~\cite{jobshop}}} 
& \JSbcmk{2}{2}{3} & 7  & 4 &    61 &   220 &   942 &   182 &    26 \\
& \JSbcmk{3}{3}{5} & 14 & 6 &    76 &   285 &  1396 &   291 &    24 \\
& \JSbcmk{3}{4}{6} & 18 & 7 &   116 &   481 &  2529 &  1241 &    21 \\
& \JSbcmk{4}{5}{7} & 27 & 9 &   141 &   606 &  2998 &  3058 &    22 \\

In [1483]:
# 选择 `culled_depth` 和 `num_params` 中的 `QTOSimplifyDiscardSolver` 列

# 查看结果
pivot_df


depth                                                    num_params  \
     HeaSolver PenaltySolver ChocoSolver QTOSimplifyDiscardSolver  HeaSolver   
pkid                                                                           
0         46.0          87.0       507.0               102.000000       90.0   
1         86.0         153.0      1532.0               306.000000      210.0   
2        121.0         174.0      2688.0              1312.000000      315.0   
3        156.0         207.0      3848.0              2910.800000      420.0   
4         76.0         259.8      1878.5               176.000000      180.0   
5         91.0         309.0      1678.5               193.000000      225.0   
6        136.0         785.4      3656.8               312.100000      360.0   
7        156.0         833.7      3263.4               256.444444      420.0   
8         56.0         175.0      1328.2               299.000000      120.0   
9        106.0         282.3      4395.3              1126.400000      270.0   
10       136.0         356.2      6137.0              3056.300000      360.0   
11       151.0         404.5      7024.4              3308.300000      405.0   
12        51.0         110.0      1062.0               215.000000      105.0   
13        86.0         170.1      3217.0               613.100000      210.0   
14       106.0         191.0      4562.0               756.800000      270.0   
15       151.0         240.0      7687.0              1815.800000      405.0   
16        61.0         221.3      1304.4               234.444444      135.0   
17        76.0         292.6      1483.6               264.000000      180.0   
18       116.0         489.8      2194.1              1084.300000      300.0   
19       141.0         627.7      2906.2              1977.000000      375.0   

                                                         \
     PenaltySolver ChocoSolver QTOSimplifyDiscardSolver   
pkid                                                      
0             10.0        10.0                 3.000000   
1             10.0        10.0                 5.000000   
2             10.0        10.0                21.000000   
3             10.0        10.0                52.000000   
4             10.0        10.0                 4.000000   
5             10.0        10.0                 3.000000   
6             10.0        10.0                 6.200000   
7             10.0        10.0                 5.222222   
8             10.0        10.0                 3.500000   
9             10.0        10.0                13.100000   
10            10.0        10.0                35.400000   
11            10.0        10.0                38.300000   
12            10.0        10.0                 3.500000   
13            10.0        10.0                10.000000   
14            10.0        10.0                11.200000   
15            10.0        10.0                24.300000   
16            10.0        10.0                 7.555556   
17            10.0        10.0                12.000000   
18            10.0        10.0                51.000000   
19            10.0        10.0               116.700000   

       culled_depth_per_param  
     QTOSimplifyDiscardSolver  
pkid                           
0                   34.000000  
1                   61.200000  
2                   62.476190  
3                   55.976923  
4                   44.000000  
5                   64.333333  
6                   50.338710  
7                   49.106383  
8                   85.428571  
9                   85.984733  
10                  86.336158  
11                  86.378590  
12                  61.428571  
13                  61.310000  
14                  67.571429  
15                  74.724280  
16                  31.029412  
17                  22.000000  
18                  21.260784  
19                  16.940874